src/storage/lock_manager/deadlock.rs

```diff
@@ -278,79 +325,196 @@ impl Scheduler {
             lock: Lock::default(),
         });
     }
+
+    fn change_role(&self, role: Role) {
+        self.notify_scheduler(Task::ChangeRole(role));
+    }
+}
+
+impl Coprocessor for Scheduler {}
+
+/// Implements observer traits for `Scheduler`.
+/// If the role of the node in the leader region changes, notifys the deadlock detector.
+///
+/// If the leader region is merged or splited in the node, the role of the node won't change.
+/// If the leader region is removed and the node is the leader, it will change to follower first.
+/// So there is no need to observe region change events.
+impl RoleObserver for Scheduler {
```

Comment:
Could we move this part into `rafstore/coprocessor`? Seems all observers are in it

Reply:
I don't like it. Maybe I will do it when separates storage as a component.

Reply:
I don't think it's a good idea too. Raftstore provides the Observer traits, and it should be implemented where it's used.